[csharp-netcore] Adding generic host library

devhl-labs opened the pull request:
# What is this?
This PR will add integration with Microsoft's [Generic Host](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0). The host manages application lifetime, dependency injection, application configuration, and much more. This library will provide extension methods for the host. This makes it easy to integrate a generated API wrapper into an existing application.

# How do I use this?
Here is sample code showing what it looks like to use this library.
```c#
public class Program
{
    public static async Task Main(string[] args)
    {
        var host = CreateHostBuilder(args).Build();
        IFooApi fooApi = host.Services.GetRequiredService<IFooApi>();
        Foo foo = await fooApi.GetFooAsync("todo");
    }

    public static IHostBuilder CreateHostBuilder(string[] args) => Host.CreateDefaultBuilder(args)
        .ConfigureApi((context, options) =>
        {
            ApiKeyToken token = new("<your token>");
            options.AddTokens(token);

            // optionally configure the token provider strategy, default is RateLimitProvider
            options.UseProvider<RateLimitProvider<ApiKeyToken>, ApiKeyToken>();

            options.ConfigureJsonOptions((jsonOptions) =>
            {
                // your custom converters if any
            });

            options.AddApiHttpClients(builder: builder => builder
                .AddRetryPolicy(2)
                .AddTimeoutPolicy(TimeSpan.FromSeconds(5))
                .AddCircuitBreakerPolicy(10, TimeSpan.FromSeconds(30))
                // any other middleware
            );
        });
}
```
# Other Changes
## api.mustache
The api.mustache template was also rewritten to take full advantage of proper dependency injection. There is also no longer a need for the Configuration class, so this was another reason to make changes in this class.
## Tokens
Each type of security protocol (like JWT and others) will generate a different class to represent the token. This class will store the token's credentials as well as a timespan. The timespan is used in client side rate limiting so we can ensure we don't get server throttled.
## RateLimitProvider
This class provides tokens to the IFooApi. Tokens will be rate limited at the client side to ensure we do not get rate limited by the server. The token timeout is defined in the token's constructor.
## TokenContainer
This is an internally used class that end users don't have to see. It merely holds the tokens in a way that is conducive for dependency injection.
## HostConfiguration
This class is used by extension methods on the generic host.
## ApiResponse\<T\> and ApiException
When do we expect exceptions when the response is not 200?
- GetFooAsyncWithHttpInfo ->  will not throw but HttpStatus is in the ApiResponse\<T\>
- GetFooAsync -> will throw
- GetFooOrDefaultAsync -> will return null

# Future Changes
- full annotation of nullable reference types
- migrate to System.Text.Json

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Blackclaws commented:
If I'm not mistaken this is mainly relevant for aspnetcore apps using api clients to access other services right? At least I don't know of any usage of GenericHost outside of aspnetcore.

I think this is in general a good idea to have, especially because DI is much more relevant in a serverside application. 

So far a quick LGTM from me, as it doesn't seem to change any existing behaviour, it just adds a new generator subclass.

I would, in general really like to clean up the netcore generator for the 6.x series of releases. We can finally cut some old baggage off and drop support for anything < Netcore 3.1

devhl-labs commented:
> If I'm not mistaken this is mainly relevant for aspnetcore apps using api clients to access other services right?
The host manages application lifetime, so it's not limited to any particular use case. It used to be called Web Host and was only for aspnet core, so maybe that is the confusion.

Blackclaws commented:
Ah, true they did actually extend the usage to console apps as well. I'm not doing a lot of csharp work outside of aspnetcore that's probably why I never used it in a different context as well.

devhl-labs commented:
@wing328  Can i get a merge?

Blackclaws commented:
Just tried it out, produces non compiling code for me.  See the attached spec
```yaml
openapi: 3.0.3
info:
  title: MangoInc Core API Specification
  description: MangoInc Core API Specification
  version: 1.0.0

paths:
  /v1/mangos:
    get:
      operationId: "getAllMangos"
      tags:
        - Mango
      summary: Returns a list of all mangos.
      description: mangos for all
      responses:
        '200': # status code
          description: A JSON array of mangos
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: '#/components/schemas/Mango'
    post:
      operationId:  "mangoCreator"
      tags:
        - Mango
      requestBody: 
        content: 
          application/json:
            schema:
              $ref: '#/components/schemas/Mango'
      responses: 
        '200':
          description: A JSON Mango
          content:
            application/json:
              schema: 
                $ref: '#/components/schemas/Mango'
  /v1/manga:
    post:
      operationId:  "mangaMan"
      tags:
        - Mango
      requestBody: 
        content: 
          image/jpeg:
            schema:
              type: string
              format: binary
      responses: 
        '200':
          description: A JSON Mango
          content:
            application/json:
              schema: 
                $ref: '#/components/schemas/Mango'

  /v1/mangaMO:
    post:
      operationId:  "mangaManMo"
      tags:
        - Mango
      requestBody: 
            content:
              multipart/form-data:
                schema:
                  type: object
                  properties:
                    orderId:
                      type: integer
                    userId:
                      type: integer
                    fileName:
                      type: string
                      format: binary
      responses: 
        '200':
          description: A JSON Mango
          content:
            application/json:
              schema: 
                $ref: '#/components/schemas/Mango'

components:
  schemas:
    Mango:
      type: object
      properties:
        id:
          type: string
          format: uuid
        weight:
          type: integer
          minimum: 0
          maximum: 100
  
    Greeting:
        type: object
        properties:
          reply:
            type: string
            maxLength: 20
            enum:
              - hi
              - howdy
 ```

With this config:
```
library: generichost
additionalProperties:
  reUseHttpClient: true
  supportsRetry: false
  validatable: false
  buildTarget: library
  operationIsAsync: true
  operationResultTask: true
  packageName: MangoInc.Core.API.Client
  packageVersion: 0.1.0
  packageTitle: MangoInc Core Client API Contract
  sourceFolder: "src"
  targetFramework: net5.0
```

Blackclaws commented on modules/openapi-generator/src/main/resources/csharp-netcore/libraries/generichost/api.mustache:
```diff
@@ -0,0 +1,324 @@
+        {{/allParams}}
+        /// <param name="cancellationToken">Cancellation Token to cancel the request.</param>
+        /// <returns>Task&lt;ApiResponse&lt;{{#returnType}}{{returnType}}{{/returnType}}{{^returnType}}object{{/returnType}}{{#nullableReferenceTypes}}?{{/nullableReferenceTypes}}&gt;&gt;</returns>
+        Task<ApiResponse<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}object{{/returnType}}{{#nullableReferenceTypes}}?{{/nullableReferenceTypes}}>> {{operationId}}WithHttpInfoAsync({{#allParams}}{{#required}}{{{dataType}}} {{paramName}}{{^-last}}, {{/-last}}{{/required}}{{^required}}{{{dataType}}} {{paramName}} = null{{^-last}}, {{/-last}}{{/required}}{{/allParams}}{{#allParams.0 ...
```
CancellationToken needs to be nullable anyway since you are assigning it to null, its not a reference type

Blackclaws commented on modules/openapi-generator/src/main/resources/csharp-netcore/libraries/generichost/api.mustache:
```diff
@@ -0,0 +1,324 @@
+        /// Initializes a new instance of the <see cref="{{classname}}"/> class.
+        /// </summary>
+        /// <returns></returns>
+        public {{classname}}(ILogger<{{classname}}> logger, HttpClient httpClient, {{#hasApiKeyMethods}}
```
Dangling comma in case all api key methods are false

devhl-labs replied:
Fixed

Blackclaws commented on modules/openapi-generator/src/main/resources/csharp-netcore/libraries/generichost/ClientUtils.mustache:
```diff
@@ -0,0 +1,337 @@
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="options"></param>
+        public static IHostBuilder Configure{{packageName}}(this IHostBuilder builder, Action<HostBuilderContext, HostConfiguration> options)
```
This doesn't work if your package name is a non trivial namespace, i.e. Company.ApiClients.Twitter or similar.

devhl-labs replied:
Replaced packageName with projectName and adding regex validation to ensure we can use it for a class name. I saw this projectName variable elsewhere, but didn't see it used in csharp. Let me know if this should change.

Blackclaws commented on modules/openapi-generator/src/main/resources/csharp-netcore/libraries/generichost/ClientUtils.mustache:
```diff
@@ -0,0 +1,337 @@
+        /// </summary>
+        /// <param name="services"></param>
+        /// <param name="options"></param>
+        public static void Add{{packageName}}(this IServiceCollection services, Action<HostConfiguration> options)
```
See Above

Blackclaws reviewed (changes requested):
Also the test project that is generated doesn't build because there is no longer a parameterless constructor.

Blackclaws commented:
I do like a lot of the general cleanup that was done in CSharpNetCoreClientCodegen, so I'd love to see the bugs fixed and this merged.

devhl-labs commented:
Wonderful!  Thanks for the help. I will take a look at the comments next week.

devhl-labs commented:
Addressed the comments above, also a few other improvements. Logging errors when the ApiResponded event is raised. Change the api token constructor to make the prefix optional, defaults to "Bearer ". I discovered the hasRequiredParams and used it to remove remove a todo.

devhl-labs commented:
@Blackclaws @wing328 
Can i get a merge or some feedback? Id like to start working on the next pr but i need this merged first.

wing328 commented:
Reviewing....

wing328 commented:
I got the following errors with a local project: 
```
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Client/HostConfiguration.cs(90,13,90,119): error CS0103: The name 'AddCmcHttpClients' does not exist in the current context
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Client/ClientUtils.cs(291,22,291,39): error CS1061: 'HostConfiguration' does not contain a definition for 'AddCmcHttpClients' and no accessible extension method 'AddCmcHttpClients' accepting a first argument of type 'HostConfiguration' could be found (are you missing a using directive or an assembly reference?)
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/FakeApi.cs(819,20,819,119): error CS0019: Operator '??' cannot be applied to operands of type 'bool' and '<throw expression>'
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/FakeApi.cs(842,19,842,23): error CS0037: Cannot convert null to 'bool' because it is a non-nullable value type
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/FakeApi.cs(1053,20,1053,119): error CS0019: Operator '??' cannot be applied to operands of type 'decimal' and '<throw expression>'
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/FakeApi.cs(1076,19,1076,23): error CS0037: Cannot convert null to 'decimal' because it is a non-nullable value type
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/PetApi.cs(513,66,513,83): error CS1503: Argument 3: cannot convert from 'System.Threading.CancellationToken?' to 'System.Threading.CancellationToken'
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/PetApi.cs(768,66,768,83): error CS1503: Argument 3: cannot convert from 'System.Threading.CancellationToken?' to 'System.Threading.CancellationToken'
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/PetApi.cs(907,66,907,83): error CS1503: Argument 3: cannot convert from 'System.Threading.CancellationToken?' to 'System.Threading.CancellationToken'
/Users/williamcheng/Code/openapi-generator2/samples/client/petstore/csharp-netcore/OpenAPIClient-generichost/src/Org.OpenAPITools/Api/PetApi.cs(1165,66,1165,83): error CS1503: Argument 3: cannot convert from 'System.Threading.CancellationToken?' to 'System.Threading.CancellationToken'
    43 Warning(s)
    10 Error(s)

```

Here is the config to generate the code:
```
generatorName: csharp-netcore
outputDir: samples/client/petstore/csharp-netcore/OpenAPIClient-generichost
inputSpec: modules/openapi-generator/src/test/resources/3_0/java/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml
templateDir: modules/openapi-generator/src/main/resources/csharp-netcore
library: generichost
additionalProperties:
  packageGuid: '{321C8C3F-0156-40C1-AE42-D59761FB9B6C}'
  useCompareNetObjects: true
  disallowAdditionalPropertiesIfNotPresent: false
  useOneOfDiscriminatorLookup: true
  targetFramework: netcoreapp3.1
```

In the IDE (VS for Mac), I've manually changed the target framework to .NET 5.0.

Does the output compile without issue in your local environment?

devhl-labs commented:
Fixed these issues. Mostly problems regarding nullable reference types. I usually have it enabled so I missed. Note that since .NET 6 NRT is enabled by default in new projects, so we should also be at least encouraging it.

wing328 commented:
@devhl-labs thanks for the quick fix. Do you need to manually change the target framework in order to make the project compiles in your environment?

> Fixing this caused issues with NRT enabled. I will take a closer look, but not until next week.

Please take your time. We'll target to include this change in the 5.4.0 release scheduled in 2022 Jan.

devhl-labs commented:
@wing328 Tested in net3.1 and net5.0 with and without NRT using the swagger file you used.

wing328 commented:
@devhl-labs thanks for the update. May I know if you've time for a quick chat via Slack?

wing328 commented:
Tested locally and the result is good. Thanks for the PR.
